features/202-session-management/config.md

```diff
@@ -0,0 +1,51 @@
+# Session Configuration
+
+## Sample configuration
```

Comment:
I think a better design is just to have multiple API keys, and each API keys can be used to request a cookie-based session / header-based session, which will be determined by the SDK so it is almost invisible to developer users.

Reply:
First half of your sentence is same as proposed design: developer can configure properties (e.g. session token transport) for each client (i.e. API key). I'm not sure what the second half (determined by the SDK) means.

Reply:
I meant, each API keys can be used for requesting a cookie-based session OR a header-based session, not pre-determined by the user configuration.

The second half meant, either requesting a cookie or header-based session don't have to be explicitly set by developers (at least for common usages), as the SDK will request the right kind of session for its usage (Web use Cookies, Mobile use header)

Reply:
I've discussed this with the team previously, and (please correct me if I'm wrong @louischan-oursky @carmenlau) the conclusion is that we don't fully understanding the security implication of this yet. There are some issues raised:
- Behavior when session token present in both header and cookie
- Session portability between front-ends
- Refresh token handling for cookies

For me, it seems dangerous. Using cookie would require awareness of CSRF prevention. Using header would expose the session token to client-side JS. Developers should determine which transport is acceptable to them.

Reply:
Please figure it out. I think requires users to pick api keys / understand all these before they start using skygear, usability-wise is unacceptable. I have yet to see any similar platform requires it when the user start using the platform for the first time.

The requirement is we only need to provide ONE API Key for users to use on both Web/Mobile platform, ideally we provide a sensible default for which transport to use on different platform.

I think we can disallow session portability between cookies/header, no refresh token for cookies, and always have header override cookies if both present?

If a strong argument is presented that it is unavoidable, let's have another issues to handle multiple api key and transports support, this is too confusing to mix up multiple api keys feature with session feature. Bottom line is we don't present two api keys by default to users

Reply:
Most auth platforms (e.g. Auth0) require developer to setup App and Clients at the beginning, and different Client ID/Secret (i.e. API Key) is used for different front-ends.

We had concluded that session portability should not be provided and refresh token for cookie is not useful, as mentioned in design doc. However, there is no clear choice on priority of header & cookie. I could look into it and see if there're some significant risk on accepting both transport.

There would not be 2 API keys by default. Instead, developers are required to setup the clients using presets (web/mobile) or customize settings when creating app.

Reply:
However, most Serverless platform also has a default API key and endpoint that can be used in "getting started situation".

Thanks for looking into the risk of accepting both transport -- if it indeed have problem, what about these changes?

1. We can cover multiple API keys in this spec, however, we adjust our Web JS SDK that it is smart enough to know whether the API key should be token based or cookies based (any idea how to do it?); So that by default we can still simply give 1 API key and users can use it in whatever clients.
2. Some settings as mentioned (forgot which comment, seems disappeared? Or maybe just I was on mobile) should be per-key setting like expiration

Reply:
If developer use Skygear as a Serverless platform, they can just deploy the services and get started; API key is not required. However, if they start using Skygear as a Auth platform, it is reasonable to expect developer to create API keys for each consuming client.

I've discussed this with the team:

- developer must choose which transport to use: they need to aware of the risks associated with each transport (e.g. CSRF for cookie, token accessible to scripts for header).
- single transport must be used: accepting/sending token in cookie may lead to security problem if services are not CSRF-aware.
- transport to use should be configured at server-side: client should not be able to request cookie if server would not accept it.

Therefore, single session token transport should be configure for each client at server-side. If we would like to simplify it, we can drop multiple API key support for now and provide a reasonable default transport (header).

Reply:
> If developer use Skygear as a Serverless platform, they can just deploy the services and get started; API key is not required.

Don't get it, if they use the Client SDK, they will need an API key right? Did I missed anything???

Ic, I think we can support multiple API keys, but let's make sure on the UX and getting start experienced it is streamlined.

As I mentioned in last message, if we go this route, can we have our client side SDK able to know weather the API key is cookie-based or token-based without extra configuration? E.g. in JS SDK, if a user copy/paste an API key into it, it will automatically use the right way to get the session.

Reply:
If the developer just want use Skygear as a hosting platform, they can choose to ignore auth gear and client SDK and just call service end-point directly. This does not require API key.

The client SDK should able to infer server configured session transport through server response, so developer just need to set the API key.

Reply:
> If the developer just want use Skygear as a hosting platform, they can choose to ignore auth gear and client SDK and just call service end-point directly. This does not require API key.

ic, I know, but it is not relevant to the problems we were discussing, maybe next time can try to avoid raising irrelevant comments which will confuse the main topic.

As discussed yesterday, we would use header token based session by default, and Web Client SDK can support both type of API token automatically.